presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java

```diff
@@ -0,0 +1,168 @@
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRewriteFilteringSemiJoinToInnerJoin(session);
+    }
+
+    @Override
+    public Result apply(FilterNode filterNode, Captures captures, Context context)
+    {
+        SemiJoinNode semiJoin = captures.get(SEMI_JOIN);
+
+        // Do not transform semi-join in context of DELETE
+        if (PlanNodeSearcher.searchFrom(semiJoin.getSource(), context.getLookup())
+                .where(node -> node instanceof TableScanNode && ((TableScanNode) node).isForDelete())
+                .matches()) {
+            return Result.empty();
+        }
+
+        Symbol semiJoinSymbol = semiJoin.getSemiJoinOutput();
+        Predicate<Expression> isSemiJoinSymbol = expression -> expression.equals(semiJoinSymbol.toSymbolReference());
+
+        List<Expression> conjuncts = extractConjuncts(filterNode.getPredicate());
+        if (conjuncts.stream().noneMatch(isSemiJoinSymbol)) {
+            return Result.empty();
+        }
+        Expression filteredPredicate = and(conjuncts.stream()
+                .filter(not(isSemiJoinSymbol))
+                .collect(toImmutableList()));
+
+        PlanNode filteringSourceDistinct = new AggregationNode(
+                context.getIdAllocator().getNextId(),
+                semiJoin.getFilteringSource(),
+                ImmutableMap.of(),
+                singleGroupingSet(ImmutableList.of(semiJoin.getFilteringSourceJoinSymbol())),
+                ImmutableList.of(),
+                SINGLE,
+                Optional.empty(),
+                Optional.empty());
+
+        Expression joinFilter = new ComparisonExpression(EQUAL, semiJoin.getSourceJoinSymbol().toSymbolReference(), semiJoin.getFilteringSourceJoinSymbol().toSymbolReference());
+
+        JoinNode innerJoin = new JoinNode(
+                semiJoin.getId(),
+                INNER,
+                semiJoin.getSource(),
+                filteringSourceDistinct,
+                ImmutableList.of(),
+                semiJoin.getSource().getOutputSymbols(),
+                ImmutableList.of(),
+                Optional.of(joinFilter),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty(),
+                semiJoin.getDynamicFilterId()
+                        .map(id -> ImmutableMap.of(id, semiJoin.getFilteringSourceJoinSymbol()))
+                        .orElse(ImmutableMap.of()),
+                Optional.empty());
+
+        ProjectNode project = new ProjectNode(
+                context.getIdAllocator().getNextId(),
+                innerJoin,
+                Assignments.builder()
+                        .putIdentities(innerJoin.getOutputSymbols())
+                        .put(semiJoinSymbol, TRUE_LITERAL)
+                        .build());
+
+        if (filteredPredicate.equals(TRUE_LITERAL)) {
```

Comment:
You probably could skip generating of `filteredPredicate` and just put entire predicate in `io.prestosql.sql.planner.plan.JoinNode#filter` (assuming true for `isSemiJoinSymbol`). That would simplify rule.

Reply:
done